---
title: "Analysis Workflow"
author: "Alexandra Cassano"
date: "`r Sys.Date()`"
output:
  html_document: 
    df_print: tibble
    highlight: tango
    theme: united
    toc: yes
    toc_float: 
        collapsed: true
  pdf_document:
    number_sections: yes
    keep_tex: yes
bibliography: references.bib
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Introduction

-   This code should be usable to analyze your bulk RNA-seq data.

-   **Please read through the document before making any changes. This is a template and will indicate where you need to make changes for your specific experiment.**

-   This R markdown file is broken down into "code chunks" that can be run individually.

    -   Downstream chunks rely on upstream chunks so running code out of order is not recommended.

    -   In R Studio, you can run a chunk by hitting the "play button (▶️)" on the top right of the chunk.

    -   You can run all upstream chunks by clicking the button that has a down arrow pointing to a green bar.

-   I've tried to include handy-dandy tables throughout this template to describe the inputs you need to provide for the functions I've written.
    The functions all also have R documentation attached to them.
    In R studio if you put your mouse cursor on the function name and hit F1 it should pop up the help information for the function.
    Alternatively, click on the function name and use the toolbar.
    Click the 'Code' tab then find 'Go to Help'.

*Built with R version 4.3.1. Most recent template changes made on September 28, 2023*

# Prepare work-space

First you have to set up the "work-space" so that your computer knows what code it will need to import and where all of the raw data is.
This will also get everything set up in the right format for downstream code.

## Install Packages

This segment of code should only be run if the packages are not already installed on whichever computer this is being run on.
In order to have this chunk run, hit the run chunk button.
If you're running chunk by chunk be sure to run this chunk the first time but do not run this code subsequent times.

[**Do not edit this chunk (just hit the run chunk button if needed)**]{style="color: red"}

```{r install_pkgs, eval=FALSE}
install.packages(c("BiocManager",
                   "tidyverse",
                   "gplots",
                   "ggplot2",
                   "ggrepel",
                   "stringi",
                   "pheatmap",
                   "stats",
                   "RColorBrewer"))

BiocManager::install(c("DESeq2",
                       "GO.db",
                       "GOstats",
                       "pathview",
                       "gage",
                       "uwot",
                       "gageData",
                       "GenomicRanges",
                       "Repitools",
                       "clusterProfiler",
                       "org.Mm.eg.db"
                       "DOSE",
                       "pathview",
                       "treemap",
                       "biomaRt",
                       "ashr"))
BiocManager::install("enrichplot")

```

## Load packages

Packages need to be loaded every time you run the code.
Run this chunk every time you open a new R session.
These will be used for data manipulation, analysis, and visualization.
Leave this chunk as `eval=TRUE`

[**Do not edit this chunk**]{style="color: red"}

```{r load_pkgs, eval=FALSE}
library(devtools)
library(roxygen2)
library(tidyverse)
library(org.Mm.eg.db)
library(utils) 
library(stringi)
library(stringr)
library(knitr)
library(rmarkdown)
library(biomaRt) 
library(DESeq2) 
library(ashr)
library(ggplot2)
library(ggrepel)
library(RColorBrewer)
library(uwot)
library(ggVennDiagram)
library(pheatmap)
library(tinytex)
library(stats)
library(enrichplot)
library(clusterProfiler)
library(DOSE)

if (!require(devtools)) { install.packages('devtools') }
devtools::install_github('JamesJoly/DGSEA')
devtools::install_github('amcassano/RNAseqAnalysisPackage') 
library(DGSEA)
library(analyzeRNA) #arguably the most important line of code, this imports all the code I wrote
library(org.Mm.eg.db)
library(ggpubr)
library(treemap)
library(pathview)
library(SPIA)


```

## Set working directory

This chunk will set the 'current working directory'.
You must change this to reflect what folder you'll be working in.
The working directory is all where all outputs will be saved.
All files being imported should also be in this folder.

Your working directory should contain:

1.  CSV file containing all of the raw read counts obtained from the fastq files

2.  CSV file containing the meta-data for your experiment

3.  CSV files with any gene sets of interest

[**Change the working directory variable (change what is in the quotes) You will need to use the complete path. The path should end with the name of the folder, you should omit the trailing '/'**]{style="color: red"}

```{r workingdirectory, eval=FALSE}
# set working directory
cwd <- "parentfolder/childfolder/yourfolder"
setwd(cwd)


# Warning: The working directory was changed to C:/Users/Lexi Cassano/Documents/GitHub/PackageTesting inside a notebook chunk. The working directory will be reset when the chunk is finished running. Use the knitr root.dir option in the setup chunk to change the working directory for notebook chunks.
#FIX THIS
```

## Data import & preparation

### Import CSV files

-   The `raw_counts` variable is set as the csv file containing the output of feature counts after using STAR to align reads to the reference genome.
    The csv file must be "cleaned up" by removing unneeded columns such as *Chr*, *start*, *end* prior to loading into R.

-   The `meta` variable is set as the csv file containing metadata about the current experiment.
    At a minimum, it must contain sample names and conditions/groups.

    -   **The condition column must be named "Group" & the sample ID column must be named "SampleID".**

    -   You can also include other information such as sac day, cell type, or anything else you might want to keep track of, let me know if you do though so I can help with some minor code adjustments.

-   Samples not of interest (in a combined experiment, irrelevant groups) should be removed from both raw counts and metadata.

-   The samples **must** be in the same order in the `meta` file as they are in `raw_counts`.

-   Pre-filtering is applied to the raw data to remove any genes that have a total count number across samples less than the user defined minimum.
    Additionally, genes that have a total count that is less than a user defined value, when the single max count is subtracted are removed.
    This is done to filter out genes that have a high read count for just one sample, but are not expressed in other samples.
    These filtering steps minimize noise and remove outliers.

*Avoid using spaces or other white space in the names of your files, use '\_' or CamelCase to separate words instead. Your file names must exactly match what you type below. Be sure to include the `.csv` file extension.*

[**Change`countsCSV` and `metadataCSV` file names**]{style="color: red"}

```{r import_data, eval=FALSE}
# import files
countsCSV <- "yourcountsfile.csv"
countsCSV <- paste(cwd, countsCSV, sep = "/")
raw_counts <- read.csv(countsCSV, header = TRUE, sep = ",")

metadataCSV <- "yourmetadatafile.csv"
metadataCSV <- paste(cwd, metadataCSV, sep = "/")
meta <- read.csv(metadataCSV, header = TRUE, sep = ",")
```

### CSV Cleanup

This code will accomplish a few things in order to get the files in the format needed for the rest of the code:

1.  Set the row names as the GeneID in the counts file

2.  Set groups as the data type `factor` :

    -   Replace whats in quotes with the names of your groups (add or remove lines as necessary). The contents of the `""` must *exactly* match the group as it is written in your metadata csv file. The order they appear is the order your groups will appear in plots and legends so fill in accordingly

3.  Apply pre-filtering to your data to retain only rows that have more than minimum counts across all samples and have a total count that is less than a user defined value, when the single max count is subtracted are removed.
    The more samples you have, the higher the number should be, the fewer samples you have the lower the number should be.
    It makes sense to set the number 300-1000.
    This is done to filter out genes that have a high read count for just one sample, but are not expressed in other samples.
    This minimizes noise and removes outliers.

[**Change the group names, count minimum, number of samples**]{style="color: red"}

```{r csv_cleanup, eval=FALSE}
# renames each row as the GeneID and removes the Geneid column
rownames(raw_counts) <- raw_counts$Geneid
raw_counts <-  dplyr::select(raw_counts, -Geneid)

# set groups as factors
meta$Group <- factor(meta$Group,
                         levels = c("group1",
                                    "group2",
                                    "group3",
                                    "group4",
                                    "group5",
                                    "group6"))

# pre-filtering
count_minimum <- 1000 
numberofsamples <- 20
raw_counts$row_max <- apply(raw_counts[,1:numberofsamples], 1, max)
raw_counts$row_sum <- apply(raw_counts[,1:numberofsamples], 1, sum)
raw_counts <- raw_counts %>%
  dplyr::filter((raw_counts$row_sum - raw_counts$row_max) > count_minimum) %>%
  dplyr::select(-c(row_max, row_sum))

# set the sample ID as the row name for the metadata 
rownames(meta) <- meta$SampleID
meta <- meta %>% dplyr::select(-c(SampleID))

```

### Rename groups (optional)

If you want to rename any groups from how they appear in your csv files (i.e. 'rej' -\> 'Acute Rejection') use the code below.

If this is needed change `eval=FALSE` to `eval=TRUE` .
Delete or add lines as needed.
Replace `group1`, `group2` etc with the name of the group you want to change.
Replace the text in quotes (`"new name 1"`) with what you want the new group name to be.
Fill in the factoring list so that the new group names replace any old ones, include unchanged group names as well

```{r rename_groups, eval=FALSE}
# rename groups 
meta$Group <- recode_factor(meta$Group, group1 = "new name 1")
meta$Group <- recode_factor(meta$Group, group2 = "New name 2")
meta$Group <- recode_factor(meta$Group, group3 = "new name 3")

# refactor groups
meta$Group <- factor(meta$Group,
                         levels = c("new name 1",
                                    "New name 2",
                                    "new name 3",
                                    "group4",
                                    "group5",
                                    "group6"))
```

## Set colors, shapes, labels for plots

-   This code will set global variables for plotting such as the color palette, labels, and shapes.
    The aesthetics dataframe is created for use in creating consistent plots throughout.
    One of the package functions has the different options for the colors and shapes in it.

-   For colors you can choose from black, blue, orange, green, purple, red, yellow, brown, white, or grey.
    Fill the `""` with your choice, add or remove rows as necessary but make sure each row ends in a `,` except the last row.
    If you want a solid symbol make the fill and outline the same color.
    If you would like an open symbol set the fill as white.

-   For shapes you can choose from 'circle', 'crossed circle', 'diamond', 'crossed diamond', 'square', 'crossed square', 'triangle up', 'triangle down', 'x', 'plus', 'asterik'.
    Fill in the `""` (same concept as outline color)

-   For each group/condition you have you will specify the shape of symbol as well as the outline and fill colors desired.
    Each should be in the same order of the groups, the code below will also print out your options and the order of your groups.

[**Do not edit this chunk, just run it.**]{style="color: red"}

```{r global_aesthetics, eval=FALSE}
# get the names of the groups to assign to the colors/shapes
sample_labels <- (meta$Group)
group_labels <- unique(sample_labels) %>% sort() %>% print()

analyzeRNA::getPalettes()
```

[**Change the colors and shapes. Make sure that the number of arguments provided match the number of conditions you have. Add or remove lines from the `anno_colors` in the below format for the number of different conditions you have.**]{style="color: red"}

```{r set_aesthetics, eval=FALSE}
plot_aes <- analyzeRNA::set_aes(group_labels, 
                    c("shape1", "shape2", "shape3"),
                    c("outline1", "outline2", "outline3"),
                    c("fill1", "fill2", "fill3"))

# for heatmaps
anno_colors <- list(
  Group = c(
    group1 = as.character(plot_aes$outlineID[1]),
    group2 = as.character(plot_aes$outlineID[2]),
    group3 = as.character(plot_aes$outlineID[3]),
    group4 = as.character(plot_aes$outlineID[4]),
    group5 = as.character(plot_aes$outlineID[5]),
    group6 = as.character(plot_aes$outlineID[6])
  )
)             
```

------------------------------------------------------------------------

# Getting results: DEGs and Normalized Counts

Now that you've got everything set up, its time to start analysis!
The DESeq2 package has functions to find differentially expressed genes, statistics associated with this, and normalized counts.

First a DESeq data object is created using `raw_counts` and `meta`, specifying that `group` is the variable of interest in the metadata table.
The `DESeq` function is run and from this output, normalized results (rlog transformed, and normalized counts).

Differential gene expression requires pairwise comparisons, however analysis using the normalized counts (UMAP, PCA, Sample Distances) use all samples at once.

## Create DESeq object and run DESeq

-   This code will construct the `DESeqDataSet` which takes the raw counts and metadata as inputs.

-   `design = ~Group` means that Group is your experimental variable, this variable expresses how the counts for each gene depend on the variables in metadata.
    If the experimental design is more complex than just group (i.e. group and cell type) then you can combine variables (i.e. `~ Group + CellType`) (but lmk if this is the case as we might need to make some slight edits to the code).

-   DESeq uses a generalized linear model and outputs a DEseq2 object, we will use this to get normalized counts, p-values, and log-2 fold changes.

[**Only change this chunk if you need to change the experiment design**]{style="color: red"}

```{r DESeq2, eval=FALSE}
dseq <- analyzeRNA::getDESeq(raw_counts, meta, ~Group)
```

## Regularized Log Transformation (rLog)

-   This code will apply a regularized log transformation to the data.
    This transforms the count data to the log2 scale in a way which minimizes differences between samples for rows with small counts, and which normalizes with respect to library size.

-   The `rlogtransformation` produces a similar variance stabilizing effect as `varianceStabilizingTransformation`.
    rlog is more robust in the case when the size factors vary widely.
    The rlog transformation takes more time to calculate than the VST function but it is more robust.

-   The transformation is useful when checking for outliers or as input for machine learning techniques such as clustering or linear discriminant analysis.

-   rlog takes as input a `DESeqDataSet` and returns a `RangedSummarizedExperiment` object.
    We will also save this same information as a data frame.
    We need both for various parts of code below.

-   `blind = FALSE` must be included in the code.
    This determines whether to blind the transformation to the experimental design.
    `blind=TRUE` should be used for comparing samples in an manner unbiased by prior information on samples (for QC).
    `blind=FALSE` should be used for transforming data for downstream analysis, where the full use of the design information should be made.
    If many of genes have large differences in counts due to the experimental design, it is important to set `blind=FALSE` for downstream analysis.

[**Do not edit this chunk**]{style="color: red"}

```{r rlog_normalization, eval=FALSE}
rlog_norm <- DESeq2::rlog(dseq, blind = FALSE)
rlog_df <- as.data.frame(SummarizedExperiment::assay(rlog_norm))
```

## Pairwise comparisons

-   Multiple pairwise comparisons can be made to determine the differentially expressed genes for any 2 given groups.
    Pairwise comparisons between 2 specified groups are saved as objects for use further on.

-   You can choose to use a Log Fold Change shrinking algorithm.
    This mitigates the effects of any single sample on the log fold change calculations, minimizing noise.
    Add `log2FCshrink = TRUE` to the function call if you want to use this.
    The calculation used is the "ashr" algorithm (@stephens2016.)

+----------------+-----------------------+------------------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Input          | Required?             | Example                      | Notes                                                                                                                                                                                                     |
+================+=======================+==============================+===========================================================================================================================================================================================================+
| `numerator`    | Required              | `"Tolerant"` `"Naive"` etc   | Must match exactly the way it's written in the DESeq object. If unsure run just the `resultsNames(dseq)` line of code.                                                                                    |
|                |                       |                              |                                                                                                                                                                                                           |
| string         |                       |                              |                                                                                                                                                                                                           |
+----------------+-----------------------+------------------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| `denominator`  | Required              | `"Tolerant"` `"Naive"` etc   | Must match exactly the way it's written in the DESeq object. If unsure run just the `resultsNames(dseq)` line of code.                                                                                    |
|                |                       |                              |                                                                                                                                                                                                           |
| string         |                       |                              |                                                                                                                                                                                                           |
+----------------+-----------------------+------------------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| `dseq_obj`     | Required              | dseq                         | This is the DESeq object we saved two chunks above. Don't change the name of it unless you also changed the name in the above chunks.                                                                     |
|                |                       |                              |                                                                                                                                                                                                           |
| DESeqObject    |                       |                              |                                                                                                                                                                                                           |
+----------------+-----------------------+------------------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| `contrast_str` | Optional;             | `"Group"`, `"Condition"` etc | This will be indicate what is the independent variable in your data. If you set things up as I described above it should be Group. If not, you may need to specify Condition or something else like that. |
|                |                       |                              |                                                                                                                                                                                                           |
| string         | defaults to `"Group"` |                              |                                                                                                                                                                                                           |
+----------------+-----------------------+------------------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| `tidy_result`  | Optional;             | `TRUE` or `FALSE`            | When set to TRUE the results will output as a nice dataframe. IDK why you'd want to change that. Odds are you'll leave it be                                                                              |
|                |                       |                              |                                                                                                                                                                                                           |
| TRUE/FALSE     | defaults to `TRUE`    |                              |                                                                                                                                                                                                           |
+----------------+-----------------------+------------------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| `log2FCshrink` | Optional;             | `TRUE` or `FALSE`            | If you've got a lot of variance going on in your data you might want to set this to TRUE. It will do some math to handle that. I don't understand the math but feel free to look it up if you care to.    |
|                |                       |                              |                                                                                                                                                                                                           |
| TRUE/FALSE     | defaults to `FALSE`   |                              |                                                                                                                                                                                                           |
+----------------+-----------------------+------------------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+

: Pairwise Comparison Function Inputs

[**Add the desired comparisons, change the names of the numerator and denominator and the names of the objects you're storing the results in.**]{style="color: red"}

```{r degs, eval=FALSE}
resultsNames(dseq)

one_vs_two <- analyzeRNA::pairwiseDEGresults(numerator = "comparisongroup",
                                denominator = "baselinegroup",
                                deseq_obj = dseq)
                                
two_vs_three <- analyzeRNA::pairwiseDEGresults(numerator = "othergroup",
                                denominator = "anotherone",
                                deseq_obj = dseq)

```

# Annotate Results

-   The initial `raw_counts` table, and therefore all DESeq results derived from it, use the *Ensembl ID* to identify each gene.
    However, this is not a useful identifier for most humans.

-   The BiomaRt package can be used to match the *Ensembl ID* with it's corresponding *MGI Symbol*, *MGI Description*, (and optionally things like the *Gene Biotype*, and *Entrez (NCBI) ID*).

-   The genemap can be altered to omit or include different attributes as needed.
    The genemap is then used along with a `deseq results object` in the `annotate_biomart()` function to add these descriptors to the results.

-   Any data frame that has *Ensembl ID*s in a column called `GeneID` can be used in this annotation function.
    This is required for using the annotation

## Make genemap

The genemap will create a data frame that includes all of the Ensembl Gene IDs in your entire data set and their corresponding MGI symbol and description.

[**Do not change (unless you want to add more things like gene biotype)**]{style="color: red"}

```{r make_genemap, eval=FALSE}
allgenes <- tibble::rownames_to_column(rlog_df, var = "GeneID")
allgenes <- select(allgenes, GeneID)
gmap <- analyzeRNA::createGenemap(allgenes,  GMattributes = c("ensembl_gene_id", "mgi_symbol", "mgi_description"))
```

## Annotate pairwise comparisons (DEGs)

+-------------------+--------------------------------------+--------------------------------------------+-------------------------------------------------------------------------------------------------------------------------------------+
| Input             | Required?                            | Example                                    | Notes                                                                                                                               |
+===================+======================================+============================================+=====================================================================================================================================+
| `res`             | Required                             | `rlog_df`, a pairwise comparison dataframe | must include 'GeneID' column with ensembl ID                                                                                        |
|                   |                                      |                                            |                                                                                                                                     |
| results dataframe |                                      |                                            |                                                                                                                                     |
+-------------------+--------------------------------------+--------------------------------------------+-------------------------------------------------------------------------------------------------------------------------------------+
| `gm`              | Required                             | `gmap`                                     | created using the chunk above                                                                                                       |
|                   |                                      |                                            |                                                                                                                                     |
| genemap           |                                      |                                            |                                                                                                                                     |
+-------------------+--------------------------------------+--------------------------------------------+-------------------------------------------------------------------------------------------------------------------------------------+
| `gmattri`         | Optional;                            | `c("MGI_Symbol",`                          | Dictates which columns from the gene map are kept after renaming and reordering. Can't include a column not present in the genemap. |
|                   |                                      |                                            |                                                                                                                                     |
| list of strings   | defaults to MGI symbol & description | `"MGI_Desc", "GeneType")`                  |                                                                                                                                     |
+-------------------+--------------------------------------+--------------------------------------------+-------------------------------------------------------------------------------------------------------------------------------------+

: Annotation Function Inputs

[**Add all of your pairwise comparisons, so all of your results get annotated**]{style="color: red"}

```{r annotate_degs, eval=FALSE}

one_vs_two <- analyzeRNA::annotate_biomart(one_vs_two, gmap) 
two_vs_three <- analyzeRNA::annotate_biomart(two_vs_three, gmap) 

rlog_df <- tibble::rownames_to_column(rlog_df, var = "GeneID")
rlog_df <- analyzeRNA::annotate_biomart(rlog_df, gmap)

```

# Filter & sort significant DEGs

-   The list of genes returned by the `results` function of the DESeq2 package contains all the genes above a 0.1 P-value.
    This is very generous filtering, especially with so many genes and so many pairwise comparisons being done.
    The results also include some genes that have uncalculated P-values or Log2 Fold Changes due to the presence of a sample that has been detected as "an extreme outlier".
    There are also results with uncalculated adjusted P-values if there is a very low mean normalized count.

-   The `signif_deg` function will filter out these genes and only keep DEGs that meet a user specified adjusted P-value threshold and log 2 fold change threshold.

    -   In addition, it will add a `Change Direction` column to indicate if a gene is up-regulated or down-regulated in the experimental group compared to the baseline group (as specified when retrieving results above).

    -   `Change Metric` is also added to each gene and the resultant data frame will be sorted according to this.
        The metric is calculated as $Change\space Metric = \log_2(Fold \space Change) *-\log(adj.\space P \space value)$.

+------------------------------+-----------------------------------------------------+
| Input                        | Required?                                           |
+==============================+=====================================================+
| `result` annotated dataframe | Required                                            |
+------------------------------+-----------------------------------------------------+
| `padj_cutoff` number         | Optional; defaults to 0.05                          |
+------------------------------+-----------------------------------------------------+
| `l2fc_cutoff` number         | Optional; defaults to 0.8 (0.8 Log2 FC = \~1.74 FC) |
+------------------------------+-----------------------------------------------------+

: Filter DEGs function inputs

[**Change threshold values (if desired), and adjust the results data frames being generated**]{style="color: red"}

```{r signifDEG, eval=FALSE}

# set thresholds for adjusted P values and log2 fold changes in determining which genes to keep as significant
padj_threshold <- 0.01
l2fc_threshold <- 1

# create ranked and filtered list of dif expr genes
ranked_OneTwo <- analyzeRNA::signif_deg(one_vs_two, padj_threshold, l2fc_threshold)
ranked_TwoThree <- analyzeRNA::signif_deg(two_vs_three, padj_threshold, l2fc_threshold)

```

# Export lists of significantly DEGs

The `export_genelist` function will save the output of the `signif_deg` function as a csv, exporting it to the current working directory.
The option to export just those genes that are up-regulated, just those that are down-regulated, or all genes is included.
**Do NOT include white space in the file name**

+--------------------+----------------------------+---------------------------------------------+------------------------------------------------------------------------------------------------------------------------------------------------+
| Input              | Required?                  | Example/Options                             | Notes                                                                                                                                          |
+:===================+============================+=============================================+================================================================================================================================================+
| `tab` data frame   | Required                   | the annotated and filtered signficiant DEGs |                                                                                                                                                |
+--------------------+----------------------------+---------------------------------------------+------------------------------------------------------------------------------------------------------------------------------------------------+
| `filename` string  | Required                   | `"TolVsNaiveSignifDEGs"`                    | If you are exporting just up or down the function will append up or down to the end of the file name. `".csv"` is not needed in the file name. |
+--------------------+----------------------------+---------------------------------------------+------------------------------------------------------------------------------------------------------------------------------------------------+
| `direction` string | Optional; defaults to both | `"up"`, `"down"`, or leave it blank         | if you use anything other than up or down it will export everything.                                                                           |
+--------------------+----------------------------+---------------------------------------------+------------------------------------------------------------------------------------------------------------------------------------------------+

: Export DEGs function inputs

[**Change the exports and the file names, feel free to export just up/down/both and not all three**]{style="color: red"}

```{r exportlists, eval=FALSE}

analyzeRNA::export_genelist(ranked_OneTwo, filename = "1vs2", direction = "up")
analyzeRNA::export_genelist(ranked_OneTwo, filename = "1vs2", direction = "down")
analyzeRNA::export_genelist(ranked_OneTwo, filename = "1vs2")

analyzeRNA::export_genelist(ranked_TwoThree, filename = "2vs3", direction = "up")
analyzeRNA::export_genelist(ranked_TwoThree, filename = "2vs3", direction = "down")
analyzeRNA::export_genelist(ranked_TwoThree, filename = "2vs3")

```

------------------------------------------------------------------------

# Data visualizations

-   For the sheer amount of data in an RNAseq experiment, one of the most useful tools is data visualizations.

-   This section includes dimensionality reduction visualization techniques as well visualizations of pairwise comparisons and visualizations of expression of specific genes of interest.

-   Visualizations that use all of the samples across conditions will use the rlog normalized data.
    Visualizations showing DEGs between two conditions will use the annotated results of pairwise comparisons.

## Sample to Sample Distances

Plotting sample distances will illustrate the overall similarity between samples.
This can be compared to what is expected by the experiment design.
Rlog transformed data will be used for this to ensure that sample distance isn't overly due to certain highly expressed genes.

[**Do not edit this chunk**]{style="color: red"}

```{r sample_distances, eval=FALSE}
analyzeRNA::sample_distances(rlog_norm, anno_colors, meta)
```

## Principal Component Analysis

-   Principal component analysis (PCA) will illustrate the differences between samples as a whole, reducing the number of dimensions.
    PCA is most useful for a smaller number of samples (i.e. what we're working with in bulk RNA seq experiments, but not in single cell sequencing experiments).

-   PC1 will be the x axis and differences in PC1 are responsible for the most variance in the data set.
    The percent of variation accounted for in PC1 will be specified in the axis label.
    PC2 is the y axis and PC2 is responsible for the second most variance.
    The percent of variation here will also be included in the axis label.

-   PCA has a few major limitations.

    -   You cannot say for certain what factor or combinations of factors are contributing to the variance.
        However, you do get very clear insight into overall similarity and differences between samples.

    -   Inclusion of a group of samples that are vastly different than the rest will limit the information you can glean from a PCA plot.
        If the difference between naive samples and any treatment condition is more significant than any other differences, then you will not be able to see other differences present.
        To further interrogate your samples I recommend removing samples like naive from your data set and running the PCA analysis on that.
        This will let you see what groupings emerge without something like naive driving \~80% of the variance.

    -   For a PCA plot to really give good information, you would hope to see the principal components account for \~50-60% of the total variance.
        Ideally, PC1 would account for \~30+% of the variance.

+-----------------------+----------------------------------------------+--------------------------------------------------------------+-----------------------------------------------------------------------------------------------------------------------------------------+
| Input                 | Required?                                    | Example                                                      | Notes                                                                                                                                   |
+=======================+==============================================+==============================================================+=========================================================================================================================================+
| `dseq_transform`      | Required                                     | `rlog_norm`                                                  | you need to use the actual object and not the data frame for this                                                                       |
|                       |                                              |                                                              |                                                                                                                                         |
| DESeqTransform object |                                              |                                                              |                                                                                                                                         |
+-----------------------+----------------------------------------------+--------------------------------------------------------------+-----------------------------------------------------------------------------------------------------------------------------------------+
| `plot_aes`            | Required                                     | `plot_aes`                                                   | use the dataframe created in the set_aesthetics chunk                                                                                   |
|                       |                                              |                                                              |                                                                                                                                         |
| dataframe             |                                              |                                                              |                                                                                                                                         |
+-----------------------+----------------------------------------------+--------------------------------------------------------------+-----------------------------------------------------------------------------------------------------------------------------------------+
| `plot_title`          | Optional;                                    | Any title for the plot you desire. Make sure it's in quotes. |                                                                                                                                         |
|                       |                                              |                                                              |                                                                                                                                         |
| string                | defaults to `"Principal Component Analysis"` |                                                              |                                                                                                                                         |
+-----------------------+----------------------------------------------+--------------------------------------------------------------+-----------------------------------------------------------------------------------------------------------------------------------------+
| `label_samples`       | Optional;                                    | `TRUE` or `FALSE`                                            | Setting to True will add labels to (some of) the data points in the plot. Most helpful when you want to be able to identify an outlier. |
|                       |                                              |                                                              |                                                                                                                                         |
| TRUE/FALSE            | defaults to `FALSE`                          |                                                              |                                                                                                                                         |
+-----------------------+----------------------------------------------+--------------------------------------------------------------+-----------------------------------------------------------------------------------------------------------------------------------------+

: PCA function inputs

[**Change the title and if you want to label the dots change it to TRUE**]{style="color: red"}

```{r PCA_analysis, eval=FALSE}
# run the function
analyzeRNA::pca_analysis(dseq_transform = rlog_norm,
             plot_aes = plot_aes,
             plot_title = "PCA",
             label_samples = FALSE)
```

## UMAP

-   UMAP is another dimensionality reduction technique.
    UMAP is better suited for a much higher number of samples (single cell vs bulk).
    Odds are you won't really use UMAP for anything here but I already wrote the code so I'm not getting rid of it.
    UMAP, unlike PCA does not provide you with the numerical amount of variance accounted for.

-   *Warning:* Every time the UMAP function is (re)run, the resulting plot will be different.
    This is due to "randomness" inherent in the UMAP algorithm.
    Either take steps to force reproducible results, or be cautious when re-running UMAP function.

    -   The result of the UMAP function is saved as an object so the plot can be displayed without re-calculating UMAP.
        It does not incorporate "true randomness" since computers are not capable of this.
        Functions achieve "pseudorandomness" by setting a number called a seed which changes every time the function is run.
        This number is generated using the system clock of the computer.

    -   The results can be forced to be reproducible by adding `batch = TRUE` to the umap() call, and by adding `set.seed(x)` where `x` is an integer you choose.

[**you can change the seed around, play with the neighbors and m Dist, change the title and toggle labels**]{style="color: red"}

```{r umapAnalysis, eval=FALSE}
set.seed(5)
rlogUMAP <- umap_analysis(dseq_transform = rlog_norm,
                          neighbors = 3,
                          m_dist = 0.5,
                          cond_list = sample_labels,
                          plot_aes = plot_aes,
                          plot_title = "rlog Normalized UMAP",
                          batch = TRUE,
                          label_samples = FALSE)

rlogUMAP
```

## Plot counts

The counts for specific genes can be plotted by group using the normalized counts.
This can be done for specific genes of interest.
Using the `plot_genecounts` function, the data can be saved to an object for plotting.

+--------------------------+--------------------------------------+-------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Input                    | Required?                            | Example                       | Notes                                                                                                                                                                                                                                                  |
+==========================+======================================+===============================+========================================================================================================================================================================================================================================================+
| `mgi`                    | Required                             | `"Tox"`                       | Must match exactly the way it's written in the MGI symbols. Check the genemap you created if you're unsure                                                                                                                                             |
|                          |                                      |                               |                                                                                                                                                                                                                                                        |
| string                   |                                      |                               |                                                                                                                                                                                                                                                        |
+--------------------------+--------------------------------------+-------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| `norm_df`                | Required                             | `rlog_df`                     | this must be the data frame version of your rlog normalized counts                                                                                                                                                                                     |
|                          |                                      |                               |                                                                                                                                                                                                                                                        |
| data frame               |                                      |                               |                                                                                                                                                                                                                                                        |
+--------------------------+--------------------------------------+-------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| `metadat`                | Required                             | `meta`                        | same metadata table used for everything                                                                                                                                                                                                                |
|                          |                                      |                               |                                                                                                                                                                                                                                                        |
| dataframe                |                                      |                               |                                                                                                                                                                                                                                                        |
+--------------------------+--------------------------------------+-------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| `plot_aes`               | Required                             | `plot_aes`                    | same aesthetics used throughout                                                                                                                                                                                                                        |
|                          |                                      |                               |                                                                                                                                                                                                                                                        |
| data frame               |                                      |                               |                                                                                                                                                                                                                                                        |
+--------------------------+--------------------------------------+-------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| `comparisons`            | Optional;                            | `list(`                       | Add as many comparisons (lists of 2 strings) as you want plotted. The line will be plotted regardless of significance but p val will only be added if its significant. The order that you write them will be the order they're plotted in (vertically) |
|                          |                                      |                               |                                                                                                                                                                                                                                                        |
| list of lists of strings | defaults to an empty list            | `c("Tolerant", "Naive"),`     |                                                                                                                                                                                                                                                        |
|                          |                                      |                               |                                                                                                                                                                                                                                                        |
|                          |                                      | `c("Tolerant", "Rejecting"),` |                                                                                                                                                                                                                                                        |
|                          |                                      |                               |                                                                                                                                                                                                                                                        |
|                          |                                      | `c("Naive", "Rejecting"))`    |                                                                                                                                                                                                                                                        |
+--------------------------+--------------------------------------+-------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| `label_samples`          | Optional;                            | `TRUE` or `FALSE`             | Change to TRUE if you would like samples labeled                                                                                                                                                                                                       |
|                          |                                      |                               |                                                                                                                                                                                                                                                        |
| TRUE/FALSE               | defaults to `FALSE`                  |                               |                                                                                                                                                                                                                                                        |
+--------------------------+--------------------------------------+-------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| `overlaps_allowed`       | Optional;                            | `10` `30` `50` etc            | This will dictate how many overlaps are tolerated when drawing the labels, make the number higher if you want to make sure more points are labeled despite being close together                                                                        |
|                          |                                      |                               |                                                                                                                                                                                                                                                        |
| number                   | defaults to `10`                     |                               |                                                                                                                                                                                                                                                        |
+--------------------------+--------------------------------------+-------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| `yaxistozero`            | Optional;                            | `TRUE` or `FALSE`             | if you set to TRUE the y axis will go all the way to 0. By default it'll zoom in around the counts with some space on either side                                                                                                                      |
|                          |                                      |                               |                                                                                                                                                                                                                                                        |
| TRUE/FALSE               | defaults to `FALSE`                  |                               |                                                                                                                                                                                                                                                        |
+--------------------------+--------------------------------------+-------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| `kw_hjust`               | Optional;                            | number 0-1                    | this dictates where the KW stat is placed, futz with it to move it where you want                                                                                                                                                                      |
|                          |                                      |                               |                                                                                                                                                                                                                                                        |
| number                   | defaults to `0.2`                    |                               |                                                                                                                                                                                                                                                        |
+--------------------------+--------------------------------------+-------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| `kw_vjust`               | Optional;                            | number 0-1                    | this dictates where the KW stat is placed, futz with it to move it where you want                                                                                                                                                                      |
|                          |                                      |                               |                                                                                                                                                                                                                                                        |
| number                   | defaults to `1`                      |                               |                                                                                                                                                                                                                                                        |
+--------------------------+--------------------------------------+-------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| `yaxis`                  | Optional;                            | any string                    | Change if you want a different Y axis label                                                                                                                                                                                                            |
|                          |                                      |                               |                                                                                                                                                                                                                                                        |
| string                   | defaults to `"rLog Normalized Reads` |                               |                                                                                                                                                                                                                                                        |
+--------------------------+--------------------------------------+-------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+

: Plot counts inputs

[**change the mgi symbol names & add as you wish. edit the optional variables as you wish and add comparisons as desired, the below are just examples**]{style="color: red"}

```{r plot_counts, eval=FALSE}

statcomps <- list(c("Naive", "Tolerant"), c("Naive", "Rejection"), c("Tolerant", "Rejection"))

plot_genecounts("Pdcd1", rlog_df, meta, plot_aes)
plot_genecounts("Tox", rlog_df, meta, plot_aes, comparisons = statcomps, yaxistozero = TRUE)


```

## Volcano Plot

-   A volcano plot will show all of the genes in the pairwise comparison results

    -   the X axis on the volcano plot is the Log2 Fold Change for each gene
    -   the Y axis is the -Log10 Adjusted P value for each gene
    -   a horizontal and vertical line will be added to the plot to indicate the threshold for significance for fold change and p value respectively

-   the plot will label significant genes with their MGI symbol (not all genes may be labeled if there are too many to plot without crowding the plot)

-   the function takes as input:

    -   results data frame for pairwise comparison, this should be annotated already, use the full data frame, not the one that has been filtered for significant genes already
    -   plot title string
    -   2 strings indicating the names of the conditions in the pairwise comparison. they should be in order so that the numerator for the DEG results is listed before the denominator
    -   *optional* change the threshold values as specified above
    -   *optional*

+----------------------------+-------------------------------+-------------------------------+------------------------------------------------------------------------------------------------------------------------------------------+
| Input                      | Required?                     | Example                       | Notes                                                                                                                                    |
+============================+===============================+===============================+==========================================================================================================================================+
| `deg_df`                   | Required                      |                               | annotated but *unfiltered* results dataframe                                                                                             |
|                            |                               |                               |                                                                                                                                          |
| dataframe                  |                               |                               |                                                                                                                                          |
+----------------------------+-------------------------------+-------------------------------+------------------------------------------------------------------------------------------------------------------------------------------+
| `plotTitle`                | Required                      | `"Condition 1 vs 2 DEGs"`     |                                                                                                                                          |
|                            |                               |                               |                                                                                                                                          |
| string                     |                               |                               |                                                                                                                                          |
+----------------------------+-------------------------------+-------------------------------+------------------------------------------------------------------------------------------------------------------------------------------+
| `cond1`                    | Required                      | `"Tolerant"`                  | The first condition (numerator) in the results data frame. Must be typed how it was above                                                |
|                            |                               |                               |                                                                                                                                          |
| string                     |                               |                               |                                                                                                                                          |
+----------------------------+-------------------------------+-------------------------------+------------------------------------------------------------------------------------------------------------------------------------------+
| `cond2`                    | Required                      | `"Rejecting"`                 | The second condition (denominator) in the results data frame                                                                             |
|                            |                               |                               |                                                                                                                                          |
| string                     |                               |                               |                                                                                                                                          |
+----------------------------+-------------------------------+-------------------------------+------------------------------------------------------------------------------------------------------------------------------------------+
| `l2fc_cutoff`              | Optional;                     |                               | where the cutoff lines will be drawn & which points will be colored                                                                      |
|                            |                               |                               |                                                                                                                                          |
| number                     | defaults to 1                 |                               |                                                                                                                                          |
+----------------------------+-------------------------------+-------------------------------+------------------------------------------------------------------------------------------------------------------------------------------+
| `pval_cutoff`              | Optional;                     |                               | where the cutoff lines will be drawn and which points will be colored                                                                    |
|                            |                               |                               |                                                                                                                                          |
| number                     | defaults to 0.01              |                               |                                                                                                                                          |
+----------------------------+-------------------------------+-------------------------------+------------------------------------------------------------------------------------------------------------------------------------------+
| `dircolors`                | Optional;                     | `c("green", "red", "gray70")` | The colors for upregulated in condition 1, downregulated in condition 1 and not significant. I recommend leaving the third as `"gray70"` |
|                            |                               |                               |                                                                                                                                          |
| list of 3 strings (colors) | defaults to red blue and gray |                               |                                                                                                                                          |
+----------------------------+-------------------------------+-------------------------------+------------------------------------------------------------------------------------------------------------------------------------------+
| `dirshapes`                | Optional;                     | `c(18, 18, 19)`               | The shapes for up in condition 1, down in condition 1 and not significant.                                                               |
|                            |                               |                               |                                                                                                                                          |
| list of 3 numbers (shapes) | defaults to circles           |                               |                                                                                                                                          |
+----------------------------+-------------------------------+-------------------------------+------------------------------------------------------------------------------------------------------------------------------------------+

: Volcano plots inputs

```{r volcano_plots, eval=FALSE}
analyzeRNA::volcano_plot(deg_df = one_vs_two, 
                         plotTitle = "One vs Two", 
                         cond1 = "One", cond2 = "Two")

analyzeRNA::volcano_plot(deg_df = two_vs_three, 
                         plotTitle = "Two vs Three", 
                         cond1 = "Two", cond2 = "Three",
                         l2fc_cutoff = 1.5, pval_cutoff = 0.05, 
                         dircolors = c("green", "red", "gray70"),
                         dirshapes = c(24, 25, 19))
```

## Gene Ontology Based Analysis

[The code here was adapted from this page.](https://hbctraining.github.io/DGE_workshop/lessons/09_functional_analysis.html)

Some info on GO terms generally: [GO Terms Overview website](https://geneontology.org/docs/ontology-documentation/ "Click for background info on GO terms generally"){.uri}

GO terms are organized into "Biological Processes" (BP), "Molecular Functions" (MF), and "Cellular Components" (CC).
You can do GSEA analysis on just one of these or a combination of them.

These analyses will use pairwise comparison results, not the rlog normalized counts.

### Over-representation Analysis

```{r over_representation, eval=FALSE}

```

### Gene Set Enrichment Analysis (GSEA)

The dotplot shows the number of genes associated with the first 50 terms (size) and the p-adjusted values for these terms (color).
This plot displays the top 50 genes by gene ratio (\# genes related to GO term / total number of sig genes), not p-adjusted value.

```{r get_gsea, eval=FALSE}




```

```{r gsea_dotplots, eval=FALSE}


```

```{r gsea_ridgeplots, eval=FALSE}

```

## Heatmaps

This code chunk will create a heatmap of the most significantly changed genes across all samples and their expression in each sample.
This visualization can easily show the effects of treatments on different genes.
It also will cluster genes.
Gene sets gotten from GSEA

```{r heatmaps}

```

### Shared Genes

```{r shared_degs}


```

### VennDiagram

```{r deg_venndiagram}

```
